### ISSUE TITLE
[Godot 4] Multiplayer code needs testing

### ISSUE THREAD

**Issue by valkyrienyanko** [Valks-Games/sankari#130]
> ℹ️ This issue is for the `godot4` branch

To test uncomment the following code in `Linker.cs` script and make sure `AutoHostJoin` and `DeveloperMode` props are enabled via inside the Godot editor inspector for the script.

![image](https://user-images.githubusercontent.com/6277739/190891353-5fc401cc-3f4d-4e68-9a68-0e9e79c00fec.png)

Last time I ran this, I got an error thus the reason why this issue is open!

**Comment by policiu** [comment]
Seems to work, you just need to copy the Newtonsoft.Json.dll to the output directory. Probably need to find a way to do that when building.

**Comment by valkyrienyanko** [comment]
This is the error I get when launching the game as host client.

```
--- Debugging process started ---
Godot Engine v4.0.beta1.mono.official.4ba934bf3 - https://godotengine.org
Vulkan API 1.2.0 - Using Vulkan Device #0: AMD - AMD Radeon RX 6600
 
[Client] Client connecting...
[Server] Server listening on port 25565
[Client] Client connected
[Server] Client connected with id: 0
[Server] Received: GameInfo
  [Error] 'Server' The type initializer for 'Sankari.Netcode.Server.ENetServer' threw an exception.\n   at Sankari.Netcode.Server.ENetServer.get_HandlePacket() in C:\Users\VALK-DESKTOP\Documents\Godot Projects\Sankari\Scripts\Netcode\Server\ENetServer.cs:line 9
     at Sankari.Netcode.Server.GameServer.Received(Peer peer, PacketReader packetReader, ClientPacketOpcode opcode) in C:\Users\VALK-DESKTOP\Documents\Godot Projects\Sankari\Scripts\Netcode\Server\GameServer.cs:line 222
     at Sankari.Netcode.Server.ENetServer.ENetThreadWorker(UInt16 port, Int32 maxClients) in C:\Users\VALK-DESKTOP\Documents\Godot Projects\Sankari\Scripts\Netcode\Server\ENetServer.cs:line 243
     at Sankari.Netcode.Server.ENetServer.<>c__DisplayClass28_0.<StartAsync>b__0() in C:\Users\VALK-DESKTOP\Documents\Godot Projects\Sankari\Scripts\Netcode\Server\ENetServer.cs:line 60
     at System.Threading.Tasks.Task`1.InnerInvoke()
     at System.Threading.Tasks.Task.<>c.<.cctor>b__272_0(Object obj)
     at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
  --- End of stack trace from previous location ---
     at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
     at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
  --- End of stack trace from previous location ---
     at Sankari.Netcode.Server.ENetServer.StartAsync(UInt16 port, Int32 maxClients, CancellationTokenSource cts) in C:\Users\VALK-DESKTOP\Documents\Godot Projects\Sankari\Scripts\Netcode\Server\ENetServer.cs:line 61
    at Scripts\Netcode\Server\ENetServer.cs:65
--- Debugging process stopped ---
```

**Comment by policiu** [comment]
Did you try placing Newtonsoft.Json.dll into .godot\mono\temp\bin\Debug ?

**Comment by valkyrienyanko** [comment]
Sorry but where do I get `Newtonsoft.Json.dll` from?

**Comment by valkyrienyanko** [comment]
I found a dll from `C:\Users\VALK-DESKTOP\.nuget\packages\newtonsoft.json\13.0.1\lib\net45`. I placed that in where you said and the error went away.

How did you know this was an error related to Newtonsoft.Json? Is there a solution that does not involve manually placing the dll file next to the games executable?

**Comment by policiu** [comment]
I was debugging use Visual Studio and the error mentioned Newtonsoft.Json. (Here is my [config](https://gist.github.com/policiu/46d9912470dc38b5e80a2df0a4206cf9))

We could try putting:
` <CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>`
into the Property group in our .csproj file. It will however copy _all_ of our referenced assemblies... even those we don't end up using.

**Comment by valkyrienyanko** [comment]
Seems like were not the only ones experiencing this problem.

https://github.com/godotengine/godot/issues/42271
https://github.com/godotengine/godot/issues/61404

The people participating in the discussion of those issues too have said to do what you said. My question is setting `CopyLocalLockFileAssemblies` to `true` bad because all the dlls are copied to the output directory? Is this a security concern or anything like that? I don't really know much about dlls other then they expose functions to other applications.

**Comment by policiu** [comment]
It's more of a space concern and not security concern since you can't call the functions w/o the .dlls to begin with.

Newtonsoft is used for a lot of reflection stuff we do.

**Comment by valkyrienyanko** [comment]
Closing as resolved df7d9354804e4d5d7d6f71d924e7451daad7fbdd
